messaging/src/main/java/org/axonframework/messaging/annotation/HandlerAttributes.java

```diff
@@ -0,0 +1,144 @@
+package org.axonframework.messaging.annotation;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Container for message handler attributes. Can store several {@link Map}s of {@link String} to {@link Object} under
+ * their own key, resembling a message handling method which serves several handling functions.
+ *
+ * @author Steven van Beelen
+ * @since 4.5
+ */
+public class HandlerAttributes {
+
+    private final Map<String, Map<String, Object>> attributes;
+
+    /**
+     * Constructs an empty handler attributes object.
+     */
+    public HandlerAttributes() {
+        this(new HashMap<>());
+    }
+
+    /**
+     * Construct a handler attributes object based on the given {@code handlerAttributes} {@link Map}.
+     *
+     * @param attributes a {@link Map} from handler type to attributes {@code Map} to base a {@link HandlerAttributes}
+     *                   on
+     */
+    public HandlerAttributes(Map<String, Map<String, Object>> attributes) {
+        this.attributes = attributes;
+    }
+
+    /**
+     * Put new {@code attributes} for the given {@code handlerType}.
+     *
+     * @param handlerType the type of handler to add {@code attributes} for
+     * @param attributes  the {@link Map} of attribute information to add for the given {@code handlerType}
+     */
+    public void put(String handlerType, Map<String, Object> attributes) {
+        this.attributes.put(handlerType, attributes);
+    }
+
+    /**
+     * Retrieve the attributes for the given {@code handlerType}. Might be {@code null} if there are no attributes
+     * present for the specified type.
+     *
+     * @param handlerType the type of handler to retrieve attributes for
+     * @return the attributes for the given {@code handlerType}
+     */
+    public Map<String, Object> get(String handlerType) {
+        return Collections.unmodifiableMap(attributes.get(handlerType));
+    }
+
+    /**
+     * Retrieve all attributes for all handler types stored in this {@link HandlerAttributes} object. The returned
+     * {@link Map} contains a {@link Map} of attribute name to attribute value, per handler type present.
+     *
+     * @return all attributes for all handler types stored in this {@link HandlerAttributes} object
+     */
+    public Map<String, Map<String, Object>> getAll() {
+        return Collections.unmodifiableMap(
+                attributes.entrySet().stream()
+                          .collect(Collectors.toMap(
+                                  Map.Entry::getKey,
+                                  entry -> Collections.unmodifiableMap(entry.getValue())
+                          ))
+        );
+    }
+
+    /**
+     * Retrieve all attributes for all handler types stored in this {@link HandlerAttributes} object in a prefixed
```

Comment:
We should be clear about whether this method does a simple "string prefix" approach, or whether the prefix approach is "dot-aware".
For example, given a key "SomeAttribute.value", will the prefix "Some" match, or not?
Also, I wonder what the value of this method is. HandlerWrappers will generally know what keys they need to interact with. If not, they can get all the attributes and simply iterate over those, ignoring whatever they don't care about. Right?

Reply:
Yep, done in the JavaDoc of the (new) `AnnotatedHandlerAttributes`.